Improve cli for calling module functions

helderco opened the pull request:
> **Warning**
> This is EXPERIMENTAL. There’s [known issues](#known-issues).

## Summary

This brings several improvements to the way you call functions from the CLI.

## Background

The current version only supports `dagger call [list | FUNCTION]`, without arguments, and simply prints whatever the result is. And `dagger shell` is implemented differently, using GraphQL queries. With the need for arguments and chaining calls, users have been relying on `dagger query` which is cumbersome to work with.

## What’s new?

### Arguments

You can now set these flags to provide arguments to function calls:
- Primitives:
    - `string` and `[]string`
    - `int` and `[]int`
    - `bool` and `[]bool`
- Dagger types:
    - `Directory`: value is the **path** in the host to the directory
    - `File`: value is the **path** in the host to the file
    - `Secret`: value is the **plaintext** which will be converted to `Secret` before sending it as an argument
    - `Container`: value is the **address** of a container image to pull from

There are no positional arguments support from the CLI’s POV. You can have positional arguments in your module’s functions (in code), but the SDK registers them with a name, which is what’s used in the API and CLI. Positional arguments are parsed as a command and `--ARG-NAME` based flags as their arguments.

### Chained calls

Functions and fields are converted to CLI sub-commands, each with their own set of flags. With the arguments support, this means you can replace many GraphQL queries with a simpler command syntax.

**Before**

```shell
$ dagger query -m github.com/shykes/daggerverse/helloWorld --focus <<EOF
{
  helloWorld {
    message
    withName(name: "Monde") {
      withGreeting(greeting: "Bonjour") {
        message
      }
    }
  }
}
EOF
✔ query [0.70s]
┃ {
┃     "helloWorld": {
┃         "message": "Hello, World!",
┃         "withName": {
┃             "withGreeting": {
┃                 "message": "Bonjour, Monde!"
┃             }
┃         }
┃     }
┃ }
• Engine: d5c749aafa87 (version devel ())
⧗ 2.40s ✔ 52 ∅ 19
```

**Alfter**

```shell
$ dagger call -m github.com/shykes/daggerverse/helloWorld \
    with-name --name=Monde \
    with-greeting --greeting=Bonjour \
    message
✔ build "dagger call" [0.44s]
├ [0.34s] loading module
├ [0.10s] loading objects
✔ running "dagger call with-name with-greeting message" [0.89s]
┃ Bonjour, Monde!
• Engine: d5c749aafa87 (version devel ())
⧗ 1.46s ✔ 35 ∅ 13
```

As with the current version of `dagger call`, you can omit the name of module from the command chain.
### Improved help

Functions and fields are converted to CLI sub-commands, with locally scoped flags, so you can add `--help` to the end of a command to get the list of available sub-commands and flags:

```console
$ dagger call -m github.com/shykes/daggerverse/helloWorld --silent with-name --help
Change the name

Usage:
  dagger call with-name [flags]
  dagger call with-name [command]

Aliases:
  with-name, WithName

Available Commands:
  message       Say hello to the world!
  with-greeting Change the greeting
  with-name     Change the name

Flags:
  -h, --help          help for with-name
      --name string

Global Flags:
      --cpuprofile string   collect CPU profile to path, and trace at path.trace
      --debug               Show more information for debugging
      --focus               Only show output for focused commands. (default true)
  -m, --mod string          Path to dagger.json config file for the module or a directory containing that file. Either local path (e.g. "/path/to/some/dir") or a git repo (e.g. "git://github.com/dagger/dagger?ref=branch?subpath=path/to/some/dir").
      --pprof string        serve HTTP pprof at this address
      --progress string     progress output format (auto, plain, tty) (default "auto")
  -s, --silent              disable terminal UI and progress output
      --workdir string      The host workdir loaded into dagger (default ".")

Use "dagger call with-name [command] --help" for more information about a command.
```

### Multiple “verbs”

All of the above capabilities have been extended to other commands besides `call`, which can do different things based on the last return type.

#### `dagger functions`

You can get an overview of a module’s objects, functions and fields:

```
$ dagger functions --silent -m github.com/shykes/daggerverse/helloWorld
object name   function name   description               return type
*HelloWorld   WithGreeting    Change the greeting       HelloWorld!
*HelloWorld   WithName        Change the name           HelloWorld!
*HelloWorld   Message         Say hello to the world!   String!
```

#### `dagger call`

Generic function call, that prints the result based on return type:

| Return type | Print                    |
| ----------- | ------------------------ |
| `Container` | `Container.stdout`       |
| `Directory` | `Directory.entries`      |
| `File`      | `File.contents`          |
| list        | Each value in a new line |
| Other       | The value                |


#### `dagger download`

It’s common to need files in your host so this command supports `Container`,  `Directory` and `File` return types, adding an `--export-path` flag, which calls:
- `Container.export(path)`
- `Directory.export(path)`
- `File.export(path)`

#### `dagger shell`

Only for `Container` return types (for now!). `dagger shell` has been changed to work like the other commands above. It adds an optional flag `--entrypoint` (it’s a [string slice](https://pkg.go.dev/github.com/spf13/pflag#StringSliceVar)) that allows you to override the default command defined in the container, which may not be something we can shell into.

If `--entrypoint` is not defined, it’ll default to the container’s default command if set, or `sh` if not.

## What’s next?

Unless there’s something major to change, I suggest we merge this asap as these improvements are already substantial compared to what’s on main and I need to hand this off to others. We can then collaborate on improvements with follow-up PRs instead of direct pushes to this branch.

## <a name="known-issues"></a>Known issues

*WIP*

shykes commented:
Amazing :)

Is it better to start bikeshedding the UI now, or in a follow-up PR?

helderco commented:
> Is it better to start bikeshedding the UI now, or in a follow-up PR?

I think in a follow-up as I need to hand this off.

helderco commented:
Just need to fix failing checks due to moving `internal/querybuilder` → `querybuilder` and the way that the Go module bootstraps, but it should be fixed #5900.

Good for review.

sipsma commented on cmd/dagger/call.go:
```diff
@@ -1,253 +1,40 @@
+	OnSelectObject: func(c *callContext, name string) (*modTypeDef, error) {
+		switch name {
+		case Container:
+			c.Select("stdout")
```
nit: in various previous iterations I think this grabbed both stdout+stderr and printed in separate blocks.

I think I *somewhat* prefer that behavior, but not a big deal, can iterate in follow ups if needed. We also talked previously about adding an api for combined stdout/stderr, so if trying to select both here throws a huge wrench into everything, it may be better to just leave as is and wait for that to exist.

sipsma commented on cmd/dagger/functions.go:
```diff
@@ -0,0 +1,546 @@
+	c.Select(fn.Name)
+
+	// TODO: Check required flags.
+	// TODO: Handle default values.
```
I'm actually not sure we do need to handle default values here ever; I *think* they get handled by our graphql server (and if they don't we should still handle setting them server-side)

helderco replied:
That's a good point. I just need to check if the flag has been changed or not because right now the flags are set with an empty value and if not set that's what will be sent here.

helderco replied:
This will be handled more gracefully in a follow up as it's already a part of an ongoing refactor.

sipsma commented on cmd/dagger/export.go:
```diff
@@ -0,0 +1,38 @@
+	OnSetFlags: func(cmd *cobra.Command, returnType *modTypeDef) error {
+		if returnType.Kind == dagger.Objectkind {
+			switch returnType.AsObject.Name {
+			case Directory, File:
```
nit: `Container` has export too (writes tarball), think we may as well add it here

helderco replied:
Yes, I was just concerned with adding one or two examples on each root command and expand later but it's now quite easy to add a few more that make sense. 👍

helderco replied:
Done.

sipsma commented on cmd/dagger/call.go:
```diff
@@ -1,253 +1,40 @@
+	Short: "Call a module's function.",
+	OnSelectObject: func(c *callContext, name string) (*modTypeDef, error) {
+		switch name {
+		case Container:
```
Not that this would make any sense, but if someone tried to call a Function that returned Secret, does that cause `dagger call` to error out (probably most logical behavior)? Haven't seen we handle that, but may be missing it.

helderco replied:
On `call` being generic, I'm not sure what the behavior should be for "unknown" (or not specifically handled) returns.

Right now the default is just `cmd.Print("%+v\n", result)` (in `OnResult`), so for a `Secret` it would show it's `SecretID`.

I’m leaning to just print by default for now and bikeshed later, unless there’s some specific we should not print.

sipsma commented on cmd/dagger/functions.go:
```diff
@@ -0,0 +1,546 @@
+	CheckReturnType func(*callContext, *modTypeDef) error
+
+	// OnResult is called when the query has completed and returned a result.
+	OnResult func(*callContext, *cobra.Command, modTypeDef, *any) error
```
nit: is `*any` needed vs. just `any`?

helderco replied:
Oh yes, def. a mistake, passed by my attention multiple times 🙂

sipsma commented on cmd/dagger/shell.go:
```diff
@@ -13,79 +13,71 @@ import (
-			return x, nil
-		default:
-			return "", fmt.Errorf("unexpected type %T", x)
+		if shellEntrypoint != nil {
+			ctr = ctr.WithExec(shellEntrypoint, dagger.ContainerWithExecOpts{
+				SkipEntrypoint: true,
```
Think this makes sense (related to discussion the other day on some PR), but if the container doesn't have an entrypoint, then we'll end up dropping into a shell for the most recent `WithExec` on the container as set in the Function.

I feel like that becomes confusing to explain to users.

I _think_ the ideal might be:
1. (for follow-up, no need in this PR) support a flag on `dagger shell` that lets you override the command to run; use that if set
1. Otherwise, if entrypoint set, use that (already done here)
1. Finally, default to running `sh`

So, if you agree, then for this PR maybe we can just add that last case of defaulting to `sh` if no entrypoint. Then a follow-up can add the first priority of using the flag if set.

sipsma replied:
Also, `DefaultArgs` fits into the ordering or priorities somewhere or another 🙂

helderco replied:
> 1. (for follow-up, no need in this PR) support a flag on dagger shell that lets you override the command to run; use that if set

That's already there with the `--entrypoint` flag, saved into the `shellEntrypoint` var when set.

Agree on all of that, here's the changes:
- If `--entrypoint` isn't set, check the container's entrypoint+defaultArgs;
- If no default command, run with `sh`;

helderco replied:
Done.

sipsma reviewed (commented):
Overall looks really really great, I love the `FuncCommand` implementation, ***much*** nicer and more organized than the previous iteration of all this 😍

helderco commented on cmd/dagger/functions.go:
```diff
@@ -0,0 +1,546 @@
+		},
+	}
+
+	if fn.Name != newCmd.Name() {
+		newCmd.Aliases = append(newCmd.Aliases, fn.Name)
+	}
```
Yes, it adds only one alternative casing per command, if it's different from the kebab case. This only shows when calling `--help` on that command:
```
$ dagger call -m github.com/shykes/daggerverse/helloWorld --silent with-name --help
Change the name

Usage:
  dagger call with-name [flags]
  dagger call with-name [command]

Aliases:
  with-name, WithName

Available Commands:
  message       Say hello to the world!
  with-greeting Change the greeting
  with-name     Change the name
```

I made it an alias to be easier because all the args parsing is delegated to cobra. If not an alias the easiest thing would be to manicure all the args to kebab case before parsing.

sipsma replied:
Okay cool, looking at the output now that seems totally fine. I previously imagined some crazy thing of every permutation of casing for every name, but in practice it looks good and is helpful to the user 👍

helderco commented on cmd/dagger/functions.go:
```diff
@@ -0,0 +1,546 @@
+			q := c.q.Bind(&response)
+
+			if err := q.Execute(ctx, c.c); err != nil {
+				return err
```
You're right! I can unwrap that particular error.

helderco commented on cmd/dagger/functions.go:
```diff
@@ -0,0 +1,546 @@
+				return fmt.Errorf("query selection: %w", err)
+			}
+
+			help, _ := cmd.Flags().GetBool("help")
```
I agree that it should be handled in the CLI. For sub-commands we can just limit to `-h`. In fact, the shorthand notation is a simple way to avoid conflict with function arguments for other cases as well, for example when returning a `Directory` and we want the `--export-path`. So we need a consistent way to add these flags for extra arguments without conflicting with the user's arguments.

helderco commented:
I placed it back on draft because there's a few details I want fix.

helderco commented:
There's a few issues here related to showing output, validation and error handling. I tried doing a few quick fixes but felt the need for more comprehensive changes and had a few ideas that felt doable in a day or two. Even though I've been making progress it's taking me longer than expected due to a bit of trial and error. Since we want to cut a release pretty soon, I'm taking a step back and fix the smallest things possible to make this usable for now.

I've already rebased locally but `dagger download` doesn't seem to be working for example, so even this is taking me longer than expected, but I hope in the next hour or so to open this up again.

helderco commented on cmd/dagger/export.go:
```diff
@@ -0,0 +1,38 @@
+	},
+	CheckReturnType: func(_ *callContext, _ *modTypeDef) error {
+		if exportPath == "" {
+			return fmt.Errorf("missing argument or return type not supported")
```
Done.

helderco commented:
Rebased and fixed a few details related to responses. Ready for merge but will leave it open in case there's time for another fix or two. We'll do follow-ups for the rest. 🙂
